### PR TITLE
Horizontal Layout

### PR DESCRIPTION
**Pull request by J3RN**
The new horizontal layout

TODO:
- [x] Get sections side-by-side
- [x] Smooth links between sections
- [x] Parallaxing background
- [x] Limit the downward scrolling to be section-specific
- [x] Fix animation on browsers other than Firefox
- [x] Optimize sections for theoretically infinite length
- [x] Fix admin section
- [x] Fix expanding blog posts outside of Chrome
- [x] Make carets look less terrible

### CODE REVIEW

**Comment by mrogman** [comment]
The code looks good as far as I can tell but the page doesn't finish loading and becomes unresponsive for me so I can't check out anything you're doing in the browser. The issue is probably on my end.

**Comment by J3RN** [comment]
@d30xy I'm sorry to hear that it doesn't load for you. If you can figure it out, I'd be happy to have your help.

**Comment by mrogman** [comment]
Ok, problem solved. However, the scroll animation still isn't working

**Comment by mrogman** [comment]
Well, I haven't identified the culprit but the animation works fine in firefox (broken in Chrome and Safari).

@J3RN What would you like to knock out next?

**Comment by J3RN** [comment]
I guess this is what I get for only testing against Firefox :frowning: Right now I'm working on the parallax scrolling (I'm just writing it myself - it's trivial enough to not require a plugin).

That means the only thing left is optimizing the sections to be forever-long. It's pretty open to interpretation, but by all mean - go for it!

**Comment by J3RN** [comment]
Oh! And don't forget to pull first, there's lots of recent changes!

**Comment by mrogman** [comment]
Wait where are you putting the parallax? Is it going in `<section class="header">` or in the background? I was thinking it would be in the background and the header section would be transparent acting as a viewing window for the parallax as you scrolled through the sections and would be full screen on the home page. If you had something else in mind I may need to go back and undo my last commit.

**Comment by J3RN** [comment]
I just finished the parallaxing. Unfortunately, I had to rollback your header shrinkage. It was supposed to look like that, as the image parallaxes in the background. Eventually we'll fix those headers up to be centered and pretty and whatnot.

Enjoy the parallaxing, though! :smile:

**Comment by mrogman** [comment]
Nice, looks good!

**Comment by mrogman** [comment]
I lost the date choosers in the admin interface and can't upload anything because of it.

screen shot 2015-04-14 at 6 19 09 pm

Are you having this issue?

**Comment by J3RN** [comment]
Bummer! I'm sad that things aren't working always working out great. My
laptop is dead and I'm in class, but I can check it out potentially when I
get home.

Also, I'm going to add these to the TODO list for this PR.

**Comment by mrogman** [comment]
No worries, it's not pressing. By all means, focus on your classwork!

**Comment by mrogman** [comment]
For the sake of awareness, the date picker issue is a problem in Firefox and Safari, but is fine in Chrome.

**Comment by mrogman** [comment]
Blog posts now display a short preview of the post content with a read more link below. Upon clicking the link the div expands to show the entire post and collapses when clicking the link again. I think this is advantageous in that the list will be condensed for easy navigation and posts of interest can be expanded as desired. Bad news is it's a little buggy at the moment and it doesn't seem to work yet in other browsers that I've tried besides Chrome.

**Comment by J3RN** [comment]
Believe it or not, that's actually exactly what I had envisioned for that section. I'll try to fix compatibility in a bit.

**Comment by d30xy** [comment]
Sweetness. I'm looking into it now. It may have something to do with this line :
```javascript
$blogInfo.css("height", slideHeight + "px");
```
It looks like there's some inconsistencies with .css() browser compatibility, but it could be something else.

**Comment by J3RN** [comment]
It looks like something weird happened with my rebase. Are your header sections smallish now as well? It's a simple fix, I just want to make sure I'm not messing things up.

**Comment by mrogman** [comment]
Yes, I'm seeing a height of 15vh (in styles.css.scss) , which is what I changed it to before you added the parallax. You said you reverted that change but I never saw it.

**Comment by J3RN** [comment]
Alright, I just made that fix. I also made the header section end right at the nav bar, so right when you begin to scroll down, it turns white and has the content.

Have you been seeing scroll bars on the individual sections? They appear to be there in Firefox. It's rather unsightly. I'm thinking we might do a `overflow: hidden;` and then use jQuery to map the scrolling action to actually scrolling.
It's totally a hack, but it gets those pesky scroll bars out of the way.

**Comment by mrogman** [comment]
Can you post a screenshot? All I see is the browser window scrollbar on the far right. But yes, kill all unsightly scrollbars!

**Comment by mrogman** [comment]
Nevermind, I think I see what you're saying. I can see the scroll bars while resizing the window (the view doesn't stay focused on the active section and the scrollbars are visible).

**Comment by J3RN** [comment]
I managed to nab this screenshot mid-transition. _Skills_

screen shot 2015-04-15 at 11 09 34 am

Basically, each section scrolls independently, and, as such, has it's own scroll bar.
**Why, you ask?** Well, some sections are longer than others. To get an idea of what the body of the page looks like, look at the original drawing in #48 and imagine a big box drawn around everything. With normal browser scrolling, you could scroll past the end of a short section, onto just empty whiteness.
**So**, what I decided to do instead is to have each section scroll independently, but still try to make it feel like it's all one large conglomerate mass.

**Comment by mrogman** [comment]
Your screenshotting skills are most impressive. I follow your logic and I'm on board with using jQuery for the scroll action.

**Comment by mrogman** [comment]
Although, if it's just a problem in Firefox, it looks like you can use css to solve the problem.

`.element::-webkit-scrollbar { width: 0 !important }` or `.element::-webkit-scrollbar { display: none !important }`

EDIT: I had a brain fart, Firefox is not webkit based...

**Comment by J3RN** [comment]
So I created a JSFiddle to show why Chrome doesn't work with the animation: [click here!!!11](https://jsfiddle.net/ZemdY/57/). To make it stop working, uncomment `overflow: hidden;` in the CSS.

Basically, Chrome pretends that content outside of the viewport in a section with `overflow: hidden;` does not exist. While the CSS3 standard seems to indicate that such behavior is totally valid, it's also totally annoying.

And so! My thoughts are three-part.
**Part 1**: Enable horizontal scrolling. It's pretty much our only option. I was also originally thinking that we could override the scroll action to scroll horizontally, but that might be problematic with respect to Part 3.
**Part 2**: Have little down-carets in each header to get to the lower actual-content sections. Like on [Thoughtbot's website](http://thoughtbot.com), but it would actually lower you _all the way_ into the section, not just most of the way.
**Part 3**: The lower sections will scroll normally (vertically - you know). The content section will scroll independent of the header section - we're essentially decoupling them. This solves the issue we had earlier with scroll bars in header sections.

**Problems**:
- Horizontal scrollbar. Ugly. Not sure what we can do about that.
- Getting back to the header section from a content section. You could click the link, of course, but not being able to scroll back up to the headers seems wonky. That said, you never scrolled _down_ to the content section, so expecting to scroll back _up_ seems a little unreasonable.

_I've probably had too much coffee this morning :coffee:_

**Comment by mrogman** [comment]
Wow, that is super annoying! Nice job getting to the bottom of it.

I have some major concerns with using the horizontal scrollbar, though. The horizonal scrollbar doesn't show in Chrome and if the navbar links don't work it might be incredibly confusing to figure out how to get to where you wan't to go. If you are using a mouse in that situation, there's really nothing you can do to move around. If you are using a trackpad, scrolling looks terrible because you can't really avoid scrolling horizontally without scrolling vertically a little bit too. I'm sure there's a way to achieve the same animation in chrome if we get creative with our hacking.

As for the carets and vertical navigation, I like your ideas. That's pretty much what I had in mind too.

**Comment by mrogman** [comment]
Setting `overflow: visible;` seems to fix the scroll animation issue in Chrome

...except the parallax is no longer working :(

**Comment by mrogman** [comment]
After playing around with a few ideas I'm thinking we should stick with the scroll animation we have now tied to the navbar links since it is effective and looks nice and it's now compatible with multiple browsers (the trick - and best practice - is putting `overflow: hidden;` only on the html tag). Unfortunately, it broke the parallax scrolling, but since it's not mission critical to using the application and navigation is, I think we should stick with the approach we have now. We should be able to get the parallax working anyway by tweaking the javascript or using a different mechanism for parallaxing (there are a few ways, including a pure css method) that we could try.

**Comment by J3RN** [comment]
Awesome work getting navigation to work! Odd that I didn't foresee a horizontal scroll bar not being displayed on most browsers anyway...

Right now I'm working on the down carets and decoupling the headers from content. After that, I'll try debugging my parallax effect.

Pure CSS parallax would be awesome! I found an [example on Codepen](http://codepen.io/keithclark/pen/JycFw). It's sort of difficult to understand, though.

**Comment by mrogman** [comment]
So I just attempted to debug your parallax javascript and it seems like this line is the problem 

``` javascript
$('#parallaxer').css('left', -(overlap * fraction));
```

The new left position is not being applied to parallaxer div. I also tried implementing left property using .animate(), but that was unsuccessful also.

The real question is why the css properties of the parallaxer div can't be updated after `overflow:hidden;` was removed from the body tag...

**Comment by J3RN** [comment]
What I've noticed is that `fraction` was always `0`. Suddenly, the `position` function started spitting out an absolute position (namely `0`), instead of the distance from the viewport to the left side of `penultimate-container`, which it used to be.

**TL;DR**, we just need to somehow get the viewport's position relative to the left side of the content.

**Comment by d30xy** [comment]
Yeah I'm seeing the same and the results don't change when clicking through the nav bar links.

**Comment by mrogman** [comment]
Browser compatibility issues with parallax and scroll animations have been fixed. Everything is gravy in Chrome, Safari, and FF!

**Comment by J3RN** [comment]
I added the down carets and lower section scrolling which we were discussing earlier. The carets are ugly, however, so I'll be fixing that up later. Functional, though!

**Comment by d30xy** [comment]
The transitions look great! I like how it scrolls back up to the top when you switch between sections.

I'm working on getting infinite scrolling applied to the blog post section to reduce initial load time when the archive becomes substantial in size. Plus I think it looks nice. 

I'm having some problems with [this plugin] (https://atmospherejs.com/alethes/pages), though, and I think the problem is that I am trying to apply it to a div, which is not supported, so I'm going to try to modify the plugin. There is a issue that references something similar [here] (https://github.com/alethes/meteor-pages/issues/10)

**Comment by mrogman** [comment]
Ok, so the expanding blog posts are working great across multiple browsers now. Also, the load more posts feature is working properly with a nice little fade-in transition (there are several built-in animations available to check out in the [momentum](https://github.com/percolatestudio/meteor-momentum) package). Right now the event is triggered by a button, which we should either make much prettier or get rid of it and use the bottom of the page as a trigger. A preload gif might not be a bad idea either.

**Comment by mrogman** [comment]
Alright guys -- One of the things I've been working on is getting the current section to remain centered in the viewport when resizing the browser window (previously the sections would expand, and the viewport did not follow the current section offset). I spun up two super hacky solutions to this problem, the latter of which I think looks much more 'natural' and is currently in use if you want to check it out. 

If you want to check out the first one, open the previous commit with `git checkout c93fb6fb0d04785f28d7154456b3f789ed954edb .`

Alright, so here's how it works:
- When the window resize event is triggered:
  - Hide all sections except the current section
  - If currently viewing content, hide the header
  - Fix the current section to the top left corner of the viewport
- When the window resizeEnd event is triggered:
  - Restore all sections and hidden headers
  - Jump to the new current section offset upon restoration

This all happens seamlessly so that it looks like the section stays in place and resizes with the window, like one would expect. After the resize is finished, the user is able to continue on browsing normally.

**Comment by mrogman** [comment]
Also, I created a responsive grid with breakpoints for the nav bar so that it resizes nicely. Only problem is that it breaks when the admin link is inserted. The problem is that each link container spans a column and when the admin link is added, there is one too many columns for the parent container. Since we are rethinking our login design (#52), I figure we can tack this issue onto it.

**Comment by J3RN** [comment]
@d30xy Nice work! Actually, we do have a grid system in place, which is [Bourbon Neat](http://neat.bourbon.io). However, if it's working, I'd say don't fix it - at least for now.

**Comment by mrogman** [comment]
Sorry, I didn't phrase that properly. I did use the bourbon neat framework. I just gave the nav bar it's own row and split it up accordingly.

**Comment by J3RN** [comment]
OK, awesome! Thanks for clearing that up!

**Comment by J3RN** [comment]
Alright, to get this presentation-ready, what I want is to get the admin section working again. Other things can be done post-merge.

**Comment by mrogman** [comment]
Cool, let's do it. Can you explain what you mean by getting the admin section working again? Do you mean the admin link or is there something wrong with the admin screen?

**Comment by J3RN** [comment]
@d30xy I was referring to the "Date is required" thing when created blog posts. You mentioned it a little while back.

**Comment by mrogman** [comment]
Oh, right. Yeah, the date chooser only shows up in chrome and I'm not entirely sure yet if there's much we can do about it. However, I have set it to insert today's date by default if not specified (or there is no date picker available) so the date field is no longer required. This will at least allow the admin to add new posts in other browsers.

**Comment by mrogman** [comment]
I wonder if we should just ditch the date field altogether and just have it insert today's date upon creation because why would you compose a blog post and assign a different date to it anyway?

**Comment by J3RN** [comment]
This is the approach that I’m leaning towards.

**Comment by mrogman** [comment]
It is done!

**Comment by mrogman** [comment]
The date field is also omitted for the events collection because it has the same name in the schema. Have we made a decision about what we are going to do with this section?

**Comment by J3RN** [comment]
@afuhrtrumpet I think we're ready to merge this.
